src/components/markdown-renderer/hooks/use-extract-first-headline.ts

```diff
@@ -53,17 +54,35 @@ export const useExtractFirstHeadline = (
   documentElement: React.RefObject<HTMLDivElement>,
   onFirstHeadingChange?: (firstHeading: string | undefined) => void
 ): (() => void) => {
-  const lastFirstHeading = useRef<string | undefined>()
+  const lastFirstHeadingContent = useRef<string | undefined>()
+  const currentFirstHeadingElement = useRef<HTMLHeadingElement | null>(null)
 
-  return useCallback(() => {
-    if (!onFirstHeadingChange || !documentElement.current) {
+  const extractHeaderText = useCallback(() => {
+    if (!onFirstHeadingChange) {
       return
     }
-    const firstHeading = documentElement.current.getElementsByTagName('h1').item(0)
-    const headingText = extractInnerText(firstHeading).trim()
-    if (headingText !== lastFirstHeading.current) {
-      lastFirstHeading.current = headingText
+    const headingText = extractInnerText(currentFirstHeadingElement.current).trim()
+    if (headingText !== lastFirstHeadingContent.current) {
+      lastFirstHeadingContent.current = headingText
       onFirstHeadingChange(headingText)
     }
-  }, [documentElement, onFirstHeadingChange])
+  }, [onFirstHeadingChange])
+
+  const mutationObserver = useMemo(() => new MutationObserver(() => extractHeaderText()), [extractHeaderText])
+  useEffect(() => () => mutationObserver.disconnect(), [mutationObserver])
+
+  return useCallback(() => {
+    const foundFirstHeading = Optional.ofNullable(documentElement.current)
+      .map((currentDocumentElement) => currentDocumentElement.getElementsByTagName('h1').item(0))
+      .orElse(null)
+    if (foundFirstHeading === currentFirstHeadingElement.current) {
+      return
+    }
+    mutationObserver.disconnect()
```

Comment:
Why do you terminate the observer here manually? Shouldn't this be done with the disposal of the `useEffect` anyway?

Reply:
At this point we have a new headline dom-element. Therefore the observer needs to stop observing the old element to avoid side effects.

Reply:
Okay, yes that is understandable.

Reply:
The effects disconnects the mutationobserver if e.g. the components gets unloaded or the callback prop changes.